[Merged by Bors] - Added API guidelines to CONTRIBUTING.md

KDecay opened the pull request:
Closes #3497.

I added the rust API guidelines (https://rust-lang.github.io/api-guidelines/about.html) to the `CONTRIBUTING.md` file. 

## Note
As noted in #3497 we should note any areas where we deliberately disagree as they arise. If we start adding these areas it might be a good idea to remove the mention of the `API guidelines` in the `CONTRIBUTING.md` file and move it to the `engine_style_guide.md`. That way we still have the connection between the `CONTRIBUTING.md` and the `API guidelines`, but we have more "space" to work with and can go into more detail about what we agree and disagree on. 

For now I would leave it this way, because it's one less click to get to the guidelines.

alice-i-cecile commented:
In terms of areas of disagreements:
- All items have a rustdoc example: probably too strong, and the examples folder is a better fit for many uses.
- Examples use ?, not try!, not unwrap: reasonable advice much of the time, but not always.
- Only smart pointers implement Deref and DerefMut: we're probably going to deliberately violate this for wrapper types due to the lack of any better ergonomic solution

These are the only ones that are important to call out IMO.

KDecay commented:
Since we already have some areas of disagreements I moved the mention to the `engine_style_guide.md` file and removed it from the `CONTRIBUTING.md` file since we already have the link to the `engine_style_guide.md` in there.

alice-i-cecile commented on .github/contributing/engine_style_guide.md:
```diff
@@ -10,3 +14,20 @@ For more advice on contributing to the engine, see the [relevant section](../../
+Some areas mentioned in the [Rust API guidelines][Rust API guidelines] we do not agree with. These areas will be expanded whenever we find something else we do not agree with, so be sure to check these from time to time.
+
+> All items have a rustdoc example
+   - This guideline is too strong and not applicable for everything inside of the Bevy game engine. Instead we should make use of the `examples` folder and add new examples that way if needed.
```
```suggestion
   - This guideline is too strong and not applicable for everything inside of the Bevy game engine. For functionality that requires more context or need a more interactive demonstration (such as rendering or input features), make use of the `examples` folder instead.
```

alice-i-cecile commented on .github/contributing/engine_style_guide.md:
```diff
@@ -10,3 +14,20 @@ For more advice on contributing to the engine, see the [relevant section](../../
+
+## Rust API guidelines
+
+As a reference for our API development we are using the [Rust API guidelines][Rust API guidelines]. These should be respected and followed except for the following areas of disagreements.
```
```suggestion
As a reference for our API development we are using the [Rust API guidelines][Rust API guidelines]. Generally, these should be followed, except for the following areas of disagreement:
```
Slightly weakened, cleaned up wording.

alice-i-cecile commented on .github/contributing/engine_style_guide.md:
```diff
@@ -10,3 +14,20 @@ For more advice on contributing to the engine, see the [relevant section](../../
+> Examples use ?, not try!, not unwrap
+   - This guideline is usually reasonable, but not always required.
+> Only smart pointers implement Deref and DerefMut
+   - Generally a good rule of thumb, but we're probably going to deliberately violate this for wrapper types due to the lack of any better ergonomic solution.
```
```suggestion
   - Generally a good rule of thumb, but we're probably going to deliberately violate this for single-element wrapper types like `Life(u32)`. The behavior is still predictable and it significantly improves ergonomics / new user comprehension.
```

KDecay commented:
Great suggestions! Thank you.

alice-i-cecile commented:
As this is a matter of "engine coding style", I'm going to leave the final call on merging this to @cart.

cart commented:
Yup I support these changes. Good job!

cart commented:
bors r+
